provider/pkg/provider/remote/copy.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2016-2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"errors"
+
+	"github.com/pulumi/pulumi-go-provider/infer"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/archive"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/asset"
+)
+
+type Copy struct{}
+
+var _ = (infer.Annotated)((*Copy)(nil))
+
+// Copy implements Annotate which allows you to attach descriptions to the Copy resource.
+func (c *Copy) Annotate(a infer.Annotator) {
+	a.Describe(&c, "Copy a local file to a remote host.")
+}
+
+type CopyInputs struct {
+	Connection   *Connection      `pulumi:"connection" provider:"secret"`
+	Triggers     *[]interface{}   `pulumi:"triggers,optional" providers:"replaceOnDelete"`
+	LocalAsset   *asset.Asset     `pulumi:"localAsset,optional"`
+	LocalArchive *archive.Archive `pulumi:"localArchive,optional"`
+	RemotePath   string           `pulumi:"remotePath"`
+}
+
+func (c *CopyInputs) Annotate(a infer.Annotator) {
+	a.Describe(&c.Connection, "The parameters with which to connect to the remote host.")
+	a.Describe(&c.Triggers, "Trigger replacements on changes to this input.")
+	a.Describe(&c.LocalAsset, "The path of the file to be copied. Only one of LocalAsset or LocalArchive can be set.")
+	a.Describe(&c.LocalArchive, "The path of the folder or archive to be copied. Only one of LocalAsset or LocalArchive can be set.")
+	a.Describe(&c.RemotePath, "The destination path in the remote host.")
+}
+
+func (c *CopyInputs) validate() error {
+	if c.LocalAsset != nil && c.LocalArchive != nil {
+		return errors.New("only one of LocalAsset or LocalArchive can be set")
+	}
+	if c.LocalAsset == nil && c.LocalArchive == nil {
+		return errors.New("either LocalAsset or LocalArchive must be set")
+	}
+	if c.LocalAsset != nil && !c.LocalAsset.IsPath() {
+		return errors.New("LocalAsset must be a file asset")
+	}
+	if c.LocalArchive != nil && !c.LocalArchive.IsPath() {
+		return errors.New("LocalArchive must be a path to a file or directory")
+	}
```

Comment:
Do we need this restriction? Can we support an arbitrary asset or archive? 

We can relax this restriction after GA if we can't get it in time.
If we can only support file-path based assets and archives we need to document that in the field descriptions explicitly:

"An asset to upload. The asset must be a path based asset."

Reply:
> I haven't fully thought through what support for non-path-based assets/archives means and was playing it safe to get started.

That makes sense. I'm OK with delaying this until after GA.

> What should the resource do with a

> * StringAsset: write the string into the remote file?

Yes

> * RemoteAsset: download first, then handle like a path asset?

That's what I would do. If we wanted to be fancier (and more efficient), we could SSH into the target system and initiate the copy from their (avoiding the relay), but that could introduce more failure types. I would stick to download then copy.

> * RemoteArchive: same

Same

> * AssetArchive: recurse through and do one of the above for each one?

Yep

Reply:
Tracked in https://github.com/pulumi/pulumi-command/issues/431